### PR TITLE
Fix for #5254

### PR DESCRIPTION
**Pull request by BaronGreenback**
**Fix1:** (#5254) The virtual interface detection routine is supposed to add the interface names defined in the list to the excluded bind list. eg. vEthernet* excludes an interfaces starting with vEthernet.

However, as the excluded interface name is not prefixed with ! it is ignored, and is instead added to the list of bind interfaces.

This PR corrects the problem by prepending the virtual interface name upon which to match with a "!" so that it will be processed correctly.

**Fix2:** When there were multiple virtual interfaces, only the first that matched was processed. All are now processed.

### CODE REVIEW

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 1]
```diff
@@ -916,7 +916,8 @@ private void InitialiseBind(NetworkConfiguration config)
                 // Add virtual machine interface names to the list of bind exclusions, so that they are auto-excluded.
                 if (config.IgnoreVirtualInterfaces)
                 {
-                    var virtualInterfaceNames = config.VirtualInterfaceNames.Split(',');
+                    // each virtual interface name must be pre-pended with the exclusion symbol !
+                    var virtualInterfaceNames = config.VirtualInterfaceNames.Split(',').Select(p => '!' + p).ToArray();
```
Comment (crobibero) [inline Jellyfin.Networking/Manager/NetworkManager.cs:920]:
```suggestion
                    var virtualInterfaceNames = config.VirtualInterfaceNames.Split(',').Select(p => "!" + p).ToArray();
```

**Review by crobibero — APPROVED** [review]

**Comment by BaronGreenback** [comment]
Fixed additional bug found whilst fixing tests.

**Comment by BaronGreenback** [comment]
Sorry for all the updates - tests worked locally, but not remotely even VS reported both were synced. Turns out they weren't!

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 2]
```diff
@@ -691,11 +691,11 @@ private void ConfigurationUpdated(object? sender, ConfigurationUpdateEventArgs e
         /// Checks the string to see if it matches any interface names.
         /// </summary>
         /// <param name="token">String to check.</param>
-        /// <param name="index">Interface index number.</param>
+        /// <param name="index">Interface index numbers that match.</param>
         /// <returns><c>true</c> if an interface name matches the token, <c>False</c> otherwise.</returns>
-        private bool IsInterface(string token, out int index)
+        private bool IsInterface(string token, out List<int>? index)
```
Comment (cvium) [inline Jellyfin.Networking/Manager/NetworkManager.cs:696]: Why a nullable list? Shouldn't it be an `IReadOnlyList<int>`? In the case of no results it should be empty. Null feels bad.

Comment (BaronGreenback) [reply Jellyfin.Networking/Manager/NetworkManager.cs]: Renamed the function to TryIsInterface and added the [NotNullWhen(true)]  attribute on the method.

Used null as its more optimised - no empty list needs to be created. Likewise with IReadOnlyList, as it's a private method and only used here)

Comment (cvium) [reply Jellyfin.Networking/Manager/NetworkManager.cs]: Enumerable. Empty is static. There's no allocation.

Comment (BaronGreenback) [reply Jellyfin.Networking/Manager/NetworkManager.cs]: Looked at the code returning an Enumerable.Empty would add a level of complexity.

The current setup mimics the TryXXXXX design whereby if the method returns _true_ then _index_ has a value, if _false_ then _index_ is _null_.

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 3]
```diff
@@ -712,13 +713,17 @@ private bool IsInterface(string token, out int index)
...
                         || (partial && interfc.StartsWith(token, true, CultureInfo.InvariantCulture)))
                     {
-                        index = interfcIndex;
-                        return true;
+                        if (index == null)
+                        {
+                            index = new List<int>();
+                        }
+
+                        index.Add(interfcIndex);
```
Comment (Bond-009) [inline Jellyfin.Networking/Manager/NetworkManager.cs:721]:
```suggestion
                        index ??= new List<int>();
                        index.Add(interfcIndex);
```

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 4]
```diff
@@ -691,11 +692,11 @@ private void ConfigurationUpdated(object? sender, ConfigurationUpdateEventArgs e
         /// Checks the string to see if it matches any interface names.
         /// </summary>
         /// <param name="token">String to check.</param>
-        /// <param name="index">Interface index number.</param>
+        /// <param name="index">Interface index numbers that match.</param>
         /// <returns><c>true</c> if an interface name matches the token, <c>False</c> otherwise.</returns>
-        private bool IsInterface(string token, out int index)
+        private bool TryIsInterface(string token, [NotNullWhen(true)] out List<int>? index)
```
Comment (Bond-009) [inline Jellyfin.Networking/Manager/NetworkManager.cs:697]:
```suggestion
        private bool TryIsInterface(string token, [MaybeNullWhen(false)] out List<int> index)
```

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: Jellyfin.Networking/Manager/NetworkManager.cs (hunk 2 above)
Comment (cvium) [reply Jellyfin.Networking/Manager/NetworkManager.cs]: The code would be a little simpler with Enumerable.Empty actually, but I get your point about mimicking other Try-function behaviours.

Could you find a better name though? `TryIsInterface` doesn't really say anything meaningful about the intent.

Comment (BaronGreenback) [reply Jellyfin.Networking/Manager/NetworkManager.cs]: Renamed to TryGetInterfaces

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 5]
```diff
@@ -730,14 +731,14 @@ private void AddToCollection(Collection<IPObject> col, string token)
         {
             // Is it the name of an interface (windows) eg, Wireless LAN adapter Wireless Network Connection 1.
             // Null check required here for automated testing.
-            if (IsInterface(token, out int index))
+            if (TryGetInterfaces(token, out var index))
```
Comment (cvium) [inline Jellyfin.Networking/Manager/NetworkManager.cs:734]: Plural?
```suggestion
            if (TryGetInterfaces(token, out var indices))
```

File: Jellyfin.Networking/Manager/NetworkManager.cs [hunk 6]
```diff
@@ -730,14 +731,14 @@ private void AddToCollection(Collection<IPObject> col, string token)
...
             {
                 _logger.LogInformation("Interface {Token} used in settings. Using its interface addresses.", token);
 
-                // Replace interface tags with the interface IP's.
+                // Replace all the interface tags with the interface IP's.
                 foreach (IPNetAddress iface in _interfaceAddresses)
                 {
-                    if (Math.Abs(iface.Tag) == index
+                    if (index.Contains(Math.Abs(iface.Tag))
```
Comment (cvium) [inline Jellyfin.Networking/Manager/NetworkManager.cs:741]:
```suggestion
                    if (indices.Contains(Math.Abs(iface.Tag))
```

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**